dsn_test.go

```diff
@@ -3,36 +3,121 @@ package sentry
 )
 
-func TestDsnParsing(t *testing.T) {
-	url := "https://username:password@domain:8888/foo/bar/23"
-	dsn, err := NewDsn(url)
-	if err != nil {
-		t.Error("expected dsn to be correctly created")
+type DsnTest struct {
+	in  string
+	dsn *Dsn   // expected value after parsing
+	url string // expected Store API URL
+}
+
+//nolint: gochecknoglobals
+var dsnTests = map[string]DsnTest{
+	"AllFields": {
+		in: "https://public:secret@domain:8888/foo/bar/42",
+		dsn: &Dsn{
+			scheme:    schemeHTTPS,
+			publicKey: "public",
+			secretKey: "secret",
+			host:      "domain",
+			port:      8888,
+			path:      "/foo/bar",
+			projectID: 42,
+		},
+		url: "https://domain:8888/foo/bar/api/42/store/",
+	},
+	"MinimalSecure": {
+		in: "https://public@domain/42",
+		dsn: &Dsn{
+			scheme:    schemeHTTPS,
+			publicKey: "public",
+			host:      "domain",
+			port:      443,
+			projectID: 42,
+		},
+		url: "https://domain/api/42/store/",
+	},
+	"MinimalInsecure": {
+		in: "http://public@domain/42",
+		dsn: &Dsn{
+			scheme:    schemeHTTP,
+			publicKey: "public",
+			host:      "domain",
+			port:      80,
+			projectID: 42,
+		},
+		url: "http://domain/api/42/store/",
+	},
+}
+
+//nolint: scopelint // false positive https://github.com/kyoh86/scopelint/issues/4
+func TestNewDsn(t *testing.T) {
+	for name, tt := range dsnTests {
+		t.Run(name, func(t *testing.T) {
+			dsn, err := NewDsn(tt.in)
+			if err != nil {
+				t.Fatalf("NewDsn() error: %q", err)
+			}
+			// Internal fields
+			if diff := cmp.Diff(tt.dsn, dsn, cmp.AllowUnexported(Dsn{})); diff != "" {
+				t.Errorf("NewDsn() mismatch (-want +got):\n%s", diff)
+			}
+			// Store API URL
+			url := dsn.StoreAPIURL().String()
+			if diff := cmp.Diff(tt.url, url); diff != "" {
+				t.Errorf("dsn.StoreAPIURL() mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+type invalidDsnTest struct {
+	in  string
+	err string // expected substring of the error
+}
+
+//nolint: gochecknoglobals
+var invalidDsnTests = map[string]invalidDsnTest{
+	"Empty":     {"", "invalid scheme"},
+	"NoScheme1": {"public:secret@:8888/42", "invalid scheme"},
+	// FIXME: NoScheme2's error message is inconsistent with NoScheme1; consider
+	// avoiding leaking errors from url.Parse.
+	"NoScheme2":     {"://public:secret@:8888/42", "missing protocol scheme"},
+	"NoPublicKey":   {"https://:secret@domain:8888/42", "empty username"},
+	"NoHost":        {"https://public:secret@:8888/42", "empty host"},
+	"NoProjectID1":  {"https://public:secret@domain:8888", "empty project id"},
+	"NoProjectID2":  {"https://public:secret@domain:8888/", "empty project id"},
+	"BadURL":        {"!@#$%^&*()", "invalid url"},
+	"BadScheme":     {"ftp://public:secret@domain:8888/1", "invalid scheme"},
+	"BadProjectID":  {"https://public:secret@domain:8888/wbvdf7^W#$", "invalid project id"},
+	"BadPort":       {"https://public:secret@domain:wat/42", "invalid port"},
+	"TrailingSlash": {"https://public:secret@domain:8888/42/", "invalid project id"},
```

Comment:
This is new and documents the current behavior.

The Sentry UI displays the DSN without trailing slashes.